src/main/drivers/sdcard_spi.c

```diff
@@ -127,22 +114,76 @@ static void sdcard_reset(void)
     }
 }
 
+
+// Called in ISR context
+// Wait until idle indicated by a read value of 0xff
+busStatus_e sdcard_callbackIdle(uint32_t arg)
+{
+    sdcard_t *sdcard = (sdcard_t *)arg;
+    extDevice_t *dev = &sdcard->dev;
+
+    uint8_t idleByte = dev->bus->curSegment->rxData[0];
+
+    if (idleByte == 0xff) {
+        return BUS_READY;
+    }
+
+    if (--sdcard->idleCount == 0) {
+        dev->bus->curSegment->rxData[0] = 0x00;
+        return BUS_ABORT;
+    }
+
+    return BUS_BUSY;
+}
+
+
+// Called in ISR context
+// Wait until idle is no longer indicated by a read value of 0xff
+busStatus_e sdcard_callbackNotIdle(uint32_t arg)
+{
+    sdcard_t *sdcard = (sdcard_t *)arg;
+    extDevice_t *dev = &sdcard->dev;
+
+    uint8_t idleByte = dev->bus->curSegment->rxData[0];
+
+    if (idleByte != 0xff) {
+        return BUS_READY;
+    }
+
+    if (sdcard->idleCount-- == 0) {
+        return BUS_ABORT;
+    }
+
+    return BUS_BUSY;
+}
+
+
 /**
  * The SD card spec requires 8 clock cycles to be sent by us on the bus after most commands so it can finish its
  * processing of that command. The easiest way for us to do this is to just wait for the bus to become idle before
  * we transmit a command, sending at least 8-bits onto the bus when we do so.
  */
 static bool sdcard_waitForIdle(int maxBytesToWait)
 {
-    while (maxBytesToWait > 0) {
-        uint8_t b = spiBusTransferByte(&sdcard.busdev, 0xFF);
-        if (b == 0xFF) {
-            return true;
-        }
-        maxBytesToWait--;
-    }
+    uint8_t idleByte;
+
+    // Note that this does not release the CS at the end of the transaction
+    busSegment_t segments[] = {
```

Comment:
Should this be `static const`?

Reply:
It doesn't need to be static as the `spiWait()` call on exit blocks and so doesn't allow this to go out of scope during the DMA. It could be const, but then the call to `spiSequence()` would barf saying we'd discarded the qualifier, so I don't think it should be.

Reply:
If it is `static const` we can avoid having all of these constant values copied onto the stack and have them used directly from flash - this is why I am bringing this up.

Reply:
I understand the potential benefit, but alas in this instance we'd be hit with the following, so this structure can't be static.

```
./src/main/drivers/sdcard_spi.c:172:15: error: initializer element is not constant
  172 |         {NULL, &idleByte, sizeof (idleByte), false, sdcard_callbackIdle},
```

Reply:
Ok, segments that contain dynamic data can not be made to live in flash obviously - for others that can I think this should be considered.

Reply:
Exploring the potential memory changes, taking the following as a baseline:

```
Linking STM32F7X2 
Memory region         Used Size  Region Size  %age Used
        ITCM_RAM:       12784 B        16 KB     78.03%
      ITCM_FLASH:          0 GB        16 KB      0.00%
ITCM_FLASH_CONFIG:          0 GB        16 KB      0.00%
     ITCM_FLASH1:          0 GB       480 KB      0.00%
      AXIM_FLASH:        2504 B        10 KB     24.45%
AXIM_FLASH_CUSTOM_DEFAULTS:           8 B         6 KB      0.13%
AXIM_FLASH_CONFIG:          0 GB        16 KB      0.00%
     AXIM_FLASH1:      447115 B       480 KB     90.97%
AXIM_FLASH_CUSTOM_DEFAULTS_EXTENDED:          0 GB         0 GB
        DTCM_RAM:       39344 B        64 KB     60.03%
           SRAM1:       73524 B       176 KB     40.80%
           SRAM2:          0 GB        16 KB      0.00%
       MEMORY_B1:          0 GB         0 GB
   text	   data	    bss	    dec	    hex	filename
 442039	   7588	 105288	 554915	  877a3	./obj/main/betaflight_STM32F7X2.elf
```

With just the change of defining the segment list as `static const` as per the following costs us a static allocation for the idleByte thus:

```
    static uint8_t idleByte;

    // Note that this does not release the CS at the end of the transaction
    static const busSegment_t segments[] = {
        {NULL, &idleByte, sizeof (idleByte), false, sdcard_callbackIdle},
        {NULL, NULL, 0, false, NULL},
    };

    sdcard.idleCount = maxBytesToWait;

    // Ensure any prior DMA has completed before continuing
    spiWaitClaim(&sdcard.dev);

    spiSequence(&sdcard.dev, (busSegment_t *)&segments[0]);
```

And the memory usage is:

```
Linking STM32F7X2 
Memory region         Used Size  Region Size  %age Used
        ITCM_RAM:       12784 B        16 KB     78.03%
      ITCM_FLASH:          0 GB        16 KB      0.00%
ITCM_FLASH_CONFIG:          0 GB        16 KB      0.00%
     ITCM_FLASH1:          0 GB       480 KB      0.00%
      AXIM_FLASH:        2504 B        10 KB     24.45%
AXIM_FLASH_CUSTOM_DEFAULTS:           8 B         6 KB      0.13%
AXIM_FLASH_CONFIG:          0 GB        16 KB      0.00%
     AXIM_FLASH1:      447139 B       480 KB     90.97%
AXIM_FLASH_CUSTOM_DEFAULTS_EXTENDED:          0 GB         0 GB
        DTCM_RAM:       39344 B        64 KB     60.03%
           SRAM1:       73532 B       176 KB     40.80%
           SRAM2:          0 GB        16 KB      0.00%
       MEMORY_B1:          0 GB         0 GB
   text	   data	    bss	    dec	    hex	filename
 442063	   7588	 105296	 554947	  877c3	./obj/main/betaflight_STM32F7X2.elf
```

So this doesn't actually get us anything as with most attempts to beat the optimiser!

Reply:
Thanks for doing the testing - interesting that the compiler seems to be making this static implicitly.